src/hotspot/share/gc/g1/g1CodeRootSet.cpp

```diff
@@ -28,89 +28,280 @@
 #include "code/nmethod.hpp"
 #include "gc/g1/g1CodeRootSet.hpp"
 #include "gc/g1/heapRegion.hpp"
+#include "memory/allocation.hpp"
 #include "oops/oop.inline.hpp"
+#include "runtime/atomic.hpp"
+#include "utilities/concurrentHashTable.inline.hpp"
+#include "utilities/concurrentHashTableTasks.inline.hpp"
 
-void G1CodeRootSet::add(nmethod* nm) {
-  assert(_is_iterating == false, "should not mutate while iterating the table");
-  bool added = false;
-  if (_table == nullptr) {
-    _table = new (mtGC) Table(SmallSize, LargeSize);
+struct G1CodeRootSetHashTableValue {
+  nmethod* _nmethod;
+
+  G1CodeRootSetHashTableValue(nmethod* nmethod) : _nmethod(nmethod) { }
+};
+
+class G1CodeRootSetHashTableConfig : public StackObj {
+public:
+  using Value = G1CodeRootSetHashTableValue;
```

Comment:
Could one use `nmethod*` here directly? Having one extra indirection/layer makes it harder to follow.